src/Symfony/Component/HttpClient/Response/CurlResponse.php

```diff
@@ -0,0 +1,270 @@
+            if (isset($this->info['url']) && $info['start_time'] / 1000 < $info['total_time']) {
+                $info['total_time'] -= $info['starttransfer_time'] ?: $info['total_time'];
+                $info['starttransfer_time'] = 0.0;
+            }
+
+            if (!\in_array(curl_getinfo($this->handle, CURLINFO_PRIVATE), ['headers', 'content'], true)) {
+                $this->finalInfo = $info;
+            }
+        }
+
+        return null !== $type ? $info[$type] ?? null : $info;
+    }
+
+    public function __destruct()
+    {
+        try {
+            if (null === $this->timeout || isset($this->info['url'])) {
+                return; // pushed response
+            }
+
+            if ('content' === $waitFor = curl_getinfo($this->handle, CURLINFO_PRIVATE)) {
+                $this->close();
+            } elseif ('headers' === $waitFor) {
+                curl_setopt($this->handle, CURLOPT_PRIVATE, 'destruct');
+            }
+
+            $this->doDestruct();
+        } finally {
+            $this->close();
+
+            // Clear local caches when the only remaining handles are about pushed responses
+            if (\count($this->multi->openHandles) === \count($this->multi->pushedResponses)) {
+                $this->multi->pushedResponses = [];
+                // Schedule DNS cache eviction for the next request
+                $this->multi->dnsCache[2] = $this->multi->dnsCache[2] ?: $this->multi->dnsCache[1];
+                $this->multi->dnsCache[1] = $this->multi->dnsCache[0] = [];
+            }
+        }
+    }
+
+    private function close(): void
+    {
+        unset($this->multi->openHandles[$this->id], $this->multi->handlesActivity[$this->id]);
+        curl_multi_remove_handle($this->multi->handle, $this->handle);
+        curl_setopt_array($this->handle, [
+            CURLOPT_PRIVATE => '',
+            CURLOPT_NOPROGRESS => true,
+            CURLOPT_PROGRESSFUNCTION => null,
+            CURLOPT_HEADERFUNCTION => null,
+            CURLOPT_WRITEFUNCTION => null,
+            CURLOPT_READFUNCTION => null,
+            CURLOPT_INFILE => null,
+        ]);
+    }
+
+    private static function schedule(self $response, array &$runningResponses): void
+    {
+        if ('' === curl_getinfo($ch = $response->handle, CURLINFO_PRIVATE)) {
+            // no-op - response already completed
+        } elseif (isset($runningResponses[$i = (int) $response->multi->handle])) {
+            $runningResponses[$i][1][$response->id] = $response;
+        } else {
+            $runningResponses[$i] = [$response->multi, [$response->id => $response]];
+        }
+    }
+
+    private static function perform(\stdClass $multi): void
+    {
+        if (self::$performing) {
+            return;
+        }
+
+        try {
+            self::$performing = true;
+            while (CURLM_CALL_MULTI_PERFORM === curl_multi_exec($multi->handle, $active));
```

Comment:
Thank you. I read libcurl's doc instead.

Reply:
Well guzzle is using exact same piece of code you used, so you are not alone here. But my another point you would realize when following php manual is this loop can be optimized with `curl_multi_select` inside

Reply:
I removed `return $this` instead.

Reply:
You should double check how curl multi works. "perform" is for emptying the queue of non-blocking operations. "select" is for filling this queue by waiting for the network - ie blocking.
Calling "select" would not be an optimization, it would break the logic. Same in Buzz and Guzzle.